src/Cardano/Wallet/BlockSyncer.hs

```diff
@@ -0,0 +1,158 @@
+                countDown counter
+
+
+_getBlock
+    :: [((Hash "BlockHeader"),Block)]
+    -> (Hash "BlockHeader")
+    -> IO Block
+_getBlock blocks headerHash =
+    case filter (\(h,_) -> h == headerHash) blocks of
+        [(_,block)] -> return block
+        _ -> error "Something wrong with the test blocks"
+
+_getEpoch
+    :: [((Hash "BlockHeader"),Block)]
+    -> Word64
+    -> IO [Block]
+_getEpoch blocks epoch = do
+    let bs =  filter (\(_,Block (BlockHeader theEpoch _ _) _) -> theEpoch == epoch) blocks
+    pure $ map snd bs
+
+_getNetworkTip
+    :: [((Hash "BlockHeader"),Block)]
+    -> IORef Int
+    -> IO ((Hash "BlockHeader"), BlockHeader)
+_getNetworkTip blocks currentPosition = do
+    index <- readIORef currentPosition
+    let (headerHash, Block blockHeader _ ) = (reverse blocks) !! index
+    pure (headerHash,blockHeader)
+
+blockSyncerBracket
+    :: IORef BlocksConsumed
+    -> BlockLayerMockup
+    -> Int
+    -> IO ()
+blockSyncerBracket blocksPersisted blockLayer period =
+    bracket
+      (return ()) --(putStrLn "setting up for ticking function")
+      (\_->return ()) --(\_ -> putStrLn "cleaning for ticking function")
+      (tickingFunction blocksPersisted blockLayer period)
+
+tickingFunction
+    :: IORef BlocksConsumed
+    -> BlockLayerMockup
+    -> Int
+    -> ()
+    -> IO ()
+tickingFunction initialState blockLayer delayPeriod _ =
+    go initialState
+    `catch`
+    (\(e :: SomeException) ->
+            let msg = "Terminating tickingFunction of BlockSyncer due to " % F.shown
+            in (putStrLn . showText) $ sformat msg e
+    )
+    where
+      go
+          :: IORef BlocksConsumed
+          -> IO ()
+      go previousState  = do
+          --putStrLn "syncing the block ..."
+          (BlocksConsumed headerHashesConsumed) <- readIORef initialState
+          (currentHeaderHash, BlockHeader _ _ previousHeaderHash) <- getNetworkTip blockLayer
+          updatedHeaderHashes <-
+              case headerHashesConsumed of
+                  [] -> pure [currentHeaderHash]
+                  lastConsumed : _ | lastConsumed == currentHeaderHash ->
+                                         pure headerHashesConsumed
+                                   | lastConsumed == previousHeaderHash ->
+                                         pure $ currentHeaderHash : headerHashesConsumed
+                                   | otherwise -> do
+                                         missingHashes <- retrieveMissingBlocks [previousHeaderHash]
+                                                          headerHashesConsumed
+                                                          blockLayer
+                                         pure $ currentHeaderHash : (missingHashes ++ headerHashesConsumed)
+          writeIORef initialState $ BlocksConsumed updatedHeaderHashes
+          threadDelay (delayPeriod*1000*1000)
+          go previousState
```

Comment:
Okay, so here's the thing. This function seems a bit too explicit and conflates several responsabilities into one. As a rule of thumb, function should do one and one thing, and try not to work at different abstraction level. If we step back a little and re-think a bit the requirements / design of such function. 

- It's about ticking / polling a source of blocks and, proceed with some action when a new block is identified.

This means that we need:

- A way to poll a source and get the latest block from a source
- An action to execute when a new block is received

We want to leave it to the caller to define such behavior, and the ticking function should just bother 
about keeping track of that latest block and calling an action when a new one is discovered.
How blocks are fetched and what action is executed matters not; So we can inject the corresponding
behaviors from the arguments:

```hs
tickingFunction 
    :: IO Block
    -> (Block -> IO ())
    -> IO ()
tickingFunction getNextBlock action = do
  -- ...
```

Then, the implementation shouldn't do much more than keeping track of the
previous block fetched, and regularly polling the chain to fetch the one,
calling the action whenever the block fetched is different from the latest
block fetched. 

There's no particular need for mocking here, because the behavior of how you get blocks doesn't matter 
for the ticking function; this incidentally makes testing way easier because you get to choose that
behavior in your testing code!

Reply:
complied. Thanks for detailed directions